File: Python/scraper.py
import requests
from bs4 import BeautifulSoup
import json

def fetch_data(url):
    response = requests.get(url)
    if response.status_code != 200:
        raise Exception(f'Failed to retrieve the page: {response.status_code}')
    soup = BeautifulSoup(response.content, 'html.parser')
    tables = soup.find_all('table')
    data = []
    for i, table in enumerate(tables):
        rows = table.find_all('tr')[1:]  # Skipping header
        day = soup.find_all('h4')[i].text.strip()  # Adjusted day fetching
        for row in rows:
            cols = row.find_all('td')
            hour = cols[1].text.strip()
            price = cols[0].text.strip()
            data.append({"hour": hour, "price": price, "day": day})
    return data

# URL of the page to be scraped
url = 'https://www.e-cena.lv/'

# Fetching data
data = fetch_data(url)

# Outputting data as JSON for inspection
print(json.dumps(data, indent=4))

File: backend/src/scraper.ts
import express from 'express';
import axios from 'axios';
import cheerio from 'cheerio';
import { optimizeActionPlan } from './optimizer';
const app = express();
const PORT = 3008;

interface DataObject {
    hour: string;
    price: number;
    day: string;
}

interface ActionEntry {
    hour: string;
    day: string;
    electricity_price: number;
    action: string;
}

app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', 'http://localhost:3000');
    res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
    next();
});

app.get('/fetch-prices', async (req, res) => {
    try {
        const url = 'https://www.e-cena.lv/';
        const response = await axios.get(url);
        const $ = cheerio.load(response.data);
        const tables = $('table');
        const data: DataObject[] = [];
        tables.each((i, table) => {
            const rows = $(table).find('tr').slice(1);  // Skipping header
            const day = $('h4').eq(i).text().trim();
            rows.each((j, row) => {
                const hour = $(row).find('td').eq(1).text().trim();
                const price = parseFloat($(row).find('td').eq(0).text().trim())
                data.push({ hour, price, day });
            });
        });
        res.json(data);
    } catch (error) {
        console.error(error);
        res.status(500).send('Server Error');
    }
});

app.post('/optimize', (req, res) => {
    const data: DataObject[] = req.body.data;
    console.log("Data sent to optimizer",data)
    if (!Array.isArray(data) || data.length === 0) {
        res.status(400).send('Invalid data');
        return;
    }
    const optimizedPlan: ActionEntry[] = optimizeActionPlan(data);
    res.json(optimizedPlan);
});


app.listen(PORT, () => {
    console.log(`Server is running on http://localhost:${PORT}`);
});
